### PR TITLE
Tokenizer redesign for better model-specific feature support

### PR DESCRIPTION
**Pull request by RdoubleA**
## Motivation

- Our tokenizers are not model specific and are rather tied to the base tokenizer, e.g., SentencePiece (SP) or TikToken (TT). These both implement `tokenize_messages` but this is not configurable for model-specific special tokens
- Both of these are highly coupled with llama2 and llama3. For example, all the special tokens for [TikToken are llama3 specific](https://github.com/pytorch/torchtune/blob/71741dfa34c9fcd6a3cae8deb56dc260a8be4ef0/torchtune/modules/tokenizers/_tiktoken.py#L83) and it is difficult to customize
- This has led to [Phi3 having its own tokenizer file](https://github.com/pytorch/torchtune/blob/71741dfa34c9fcd6a3cae8deb56dc260a8be4ef0/torchtune/models/phi3/_sentencepiece.py#L35) that reimplements SentencePieceTokenizer, but with custom special tokens. Ideally, this should be easily composable by configuring the phi3 specific tokens with SentencePiece as the base
- This problem will only get worse as foundation models add more capabilities such as tool calling to their respective tokenizers. For example, [Mistral v3 now uses their own special tokens on top of SentencePiece](https://docs.google.com/document/d/1zAa956bEz-CR6ca2BnDntwA1iB3qltU4SqH9GyoToaM/edit?usp=sharing), and we have no way to onboard this besides copying and reimplementing SentencePieceTokenizer. Redundancy is normally ok but not if it can be easily addressed with composability.
- If a new base tokenizer is used besides SentencePiece or TikToken, we should be flexible enough to quickly implement the underlying new encode and decode while still maintaining our higher level APIs (tokenize_messages, etc)

## Design proposal
Intuition: separate the two core APIs (encode/decode and tokenize_messages) that operate at different levels of abstraction:
- Encode/decode live in the base token encoding layer (SP or TT) and does not need model specific logic. This should be pulled out so any model can quickly use SP, TT, or any other base token encoding for encode/decode.
- tokenize_messages requires model specific special tokens placed in particular locations. Thus, there should be a model specific class that implements tokenize_messages

We can achieve the above with `BaseTokenizer` and `ModelTokenizer`.
- `BaseTokenizer` is the base abstract interface for any base tokenization model (SP or TT) that implements encode and decode
- `ModelTokenizer` is the base abstract interface for any model-specific tokenizer that implements tokenize_messages. All models will implement their own Tokenizer class based on this interface so they can control `tokenize_messages` logic

```

class BaseTokenizer(Protocol):
    """Abstract token encoding model"""

    def encode(self, text: str, **kwargs) -> List[int]:
        """
        Given a string, return the a list of token ids.
        """

    def decode(
        self, token_ids: List[int], include_special: bool = False, **kwargs
    ) -> str:
        """
        Given a list of token ids, return the decoded text.
        """
```
```
class ModelTokenizer(Protocol):
    """
    Abstract tokenizer that implements model specific special token logic in
    ``tokenize_message`` and ``tokenize_messages`` methods.
    """

    def tokenize_messages(
        self, messages: List[Message], **kwargs
    ) -> Tuple[List[int], List[bool]]:
        """
        Given a list of messages, return a list of tokens and list of masks for
        the concatenated and formatted messages.
        """
        pass
```

This means the SentencePieceTokenizer and TikTokenTokenizer will be refactored to separate out encode/decode and tokenize_messages logic.

```
class TikTokenBaseTokenizer(BaseTokenizer):
    def __init__(
        self,
        path: str,
        pattern: str,
        name: str,
        special_tokens: Dict[str, int],
    ):
        mergeable_ranks = load_tiktoken_bpe(path)
        self.tt_model = Encoding(
            name=name,
            pat_str=pattern,
            mergeable_ranks=mergeable_ranks,
            special_tokens=special_tokens,
        )
        ...

    def encode(
        self,
        text: str,
        add_bos: bool,
        add_eos: bool,
        allowed_special: Set[str] = set(),
    ) -> List[int]:
        ...

    def decode(
        self,
        token_ids: List[int],
        truncate_at_eos: bool = True,
    ) -> str:
        ...
```
```
class SentencePieceBaseTokenizer(BaseTokenizer):
    def __init__(
        self,
        path: str,
    ):
        spm_model = SentencePieceProcessor()
        spm_model.load(path)
        self.spm_model = spm_model
        # Special tokens are defined in the protobuf itself

    def encode(
        self,
        text: str,
        add_bos: bool,
        add_eos: bool,
        allowed_special: Set[str] = set(),
        trim_leading_whitespace: bool = False,
        prefix: Optional[str] = None,
    ) -> List[int]:
        ...

    def decode(
        self,
        token_ids: List[int],
    ) -> str:
        ...
```
And any model tokenizers would compose with the above classes:
```
class Llama3Tokenizer(ModelTokenizer):
    """
    tiktoken tokenizer configured with Llama3 Instruct's special tokens, as described in
    https://llama.meta.com/docs/model-cards-and-prompt-formats/meta-llama-3

    Args:
        path (str): Path to pretrained tiktoken tokenizer file.
    """

    def __init__(
        self,
        path: str,
    ):
        all_special_tokens_with_ids = self._get_all_special_tokens_with_ids()
        self.tt_model = TikTokenBaseTokenizer(
            path=path,
            name="llama3_tiktoken",
            pattern=CL100K_PATTERN,
            special_tokens=all_special_tokens_with_ids,
        )

    def tokenize_messages(
        self,
        messages: List[Message],
        max_seq_len: Optional[int] = None,
        tokenize_header: bool = True,
        add_eos: bool = True,
    ) -> Tuple[List[int], List[bool]]:
        ...
```
## Changelog
- Refactor SentencePieceTokenizer -> SentencePieceBaseTokenizer, TikTokenTokenizer -> TikTokenBaseTokenizer and only retain encode/decode logic
- Move tokenize_messages logic to `Llama2Tokenizer` and `Llama3Tokenizer` for SP and TT, respectively
- Create `GemmaTokenizer` and `MistralTokenizer` which leverage `SentencePieceBaseTokenizer`. The `tokenize_messages` logic is identical to `Llama2Tokenizer` (for now, Mistral needs to be updated with v3)
- Refactor Phi3's tokenizer to use `SentencePieceBaseTokenizer` and retain special token logic
- Revamp `DummyTokenizer` since it inherited previously from `SentencePieceTokenizer`
- Update all tests and docstrings

## Test plan
- Added unit tests for all tokenizers, `pytest tests` runs successfully
- Manual comparison of every updated tokenizer with the old version: https://gist.github.com/RdoubleA/a62855ae5440063d120de3c908c903e0

## Planned follow-ups
- Tokenizer live docs / API ref updates

### CODE REVIEW

File: torchtune/models/llama3/_special_tokens.json
Comment (joecummings) [inline torchtune/models/llama3/_special_tokens.json:1]: I like this format a lot b/c it mirrors how tokenizers are packaged with HF - can you explicitly call this a `_special_tokens_map.json` though?

File: torchtune/models/llama3/_tokenizer.py
Comment (joecummings) [inline torchtune/models/llama3/_tokenizer.py:57]: I would prefer this to be configurable with a `file_path` param rather than hardcoded in.

File: torchtune/models/llama3/_special_tokens.json
Comment (joecummings) [reply torchtune/models/llama3/_special_tokens.json:1]: Hmm, but should this technically be included in our repo? It gets packaged with the tokenizer we download so maybe we should just point to that one instead??

Comment (joecummings) [reply torchtune/models/llama3/_special_tokens.json:1]: Thus, this file would be deleted.

Comment (RdoubleA) [reply torchtune/models/llama3/_special_tokens.json:1]: That's an interesting idea, I'm just worried that gives us less control over when certain special tokens are officially supported in our library. For example, Mistral v3 has an updated special_tokens_map but we currently only have one MistralTokenizer class that doesn't yet use the new special tokens. If we loaded in the special_tokens_map.json from HF for Mistral v3 the tokenizer will not work as expected. Then we would basically need a separate tokenzier class for each iteration of the model. We could compare it against the HF special_tokens_map.json as part of our correctness check.

Comment (joecummings) [reply torchtune/models/llama3/_special_tokens.json:1]: I think this is the same as our guarantees around supported models in torchtune. We make no claims to support Mistral V3 ATM so users should not expect that it works OOTB, despite likely being able to load it with a little messing around. The same principle applies for tokenizers.

Comment (joecummings) [reply torchtune/models/llama3/_special_tokens.json:1]: And when we make claims that we fully support a model, then that means we should fully support all the special tokens needed to run that model. Hence, the special_tokens_map file that comes with the model should be guaranteed to work with torchtune.

Comment (joecummings) [reply torchtune/models/llama3/_special_tokens.json:1]: In addition, let's say someone wants to add a special token for their finetuned model. If we allow people to point to a file with the special tokens map supplied, it will be trivial to do so.

Comment (RdoubleA) [reply torchtune/models/llama3/_special_tokens.json:1]: OK, I am mostly convinced. I looked at the jsons that are typically on the model pages on HF and `tokenizer.json` is the one we want. The others don't seem to have the complete list of special tokens. The only other thing is we'd have to add an additional field for the path to the special tokens json for all tokenizers. I do like the ease of seeing all the special tokens just listed there on our repo though, but maybe we can just document it somewhere for each model...

**Review by kartikayk — COMMENTED** [review]
Want to express my uncertainty on the ```TokenEncoding``` naming. This seems a bit unintuitive since I'd imagine the encoding to refer to something like UTF8.

Not the best, but should we consider Tokenizer (and so SentencePieceTokenizer derives from Tokenizer) and ModelTokenizer (and so Llama3Tokenizer derives from ModelTokenizer). I don't love ModelTokenizer, but with the right doc strings I think it's passable.

File: torchtune/data/tokenizers/_tiktoken.py
Comment (ebsmothers) [inline torchtune/data/tokenizers/_tiktoken.py:25]: This is a bit confusing to me (I understand the motivation though). In practice I would strive for more consistency in the `encode` APIs -- obviously there are some pretty custom things in sentencepiece that we don't wanna recreate  across the board, so not saying the signatures should match exactly or anything. But BOS and EOS handling should imo be exposed in the same place

File: torchtune/data/tokenizers/_tiktoken.py
Comment (ebsmothers) [inline torchtune/data/tokenizers/_tiktoken.py:54]: kind of a nit but strictly speaking this does not need to be a method on the class, right?

File: torchtune/models/llama3/_special_tokens.json
Comment (ebsmothers) [reply torchtune/models/llama3/_special_tokens.json:1]: A question from elsewhere on this PR that may be relevant to this discussion.. can we just define `special_tokens` as a property of the `Tokenizer` protocol? Then we have some formal contract about where it's exposed. (I agree that we should probably not store any actual JSON files like this in the repo)

**Comment by ebsmothers** [comment]
Overall I like the proposal. Still going through all the code, but the division of `encode` and `tokenize_messages` APIs you've provided here definitely makes sense to me. One big question I have is whether we can do this in a way where we don't have to write an entirely new tokenizer class every time we onboard a new model. I get that if there's custom logic around special tokens we can't really avoid it, but for vanilla sentencepiece without much special formatting it'd be nice to make this process a bit more lightweight (e.g. Gemma and Llama2 tokenizers have the same `tokenize_messages` API, and imo copy-pasting this code is not as nice as than copy-pasting model builders, which are generally pretty straightforward to understand).


Also: 

> Move torchtune.modules.tokenizers to torchtune.data.tokenizers which is a more relevant location

I understand that this is a useful thing to do, but it makes reviewing this code harder and is kinda logically distinct from the important stuff in this PR (imo). With this move, I have to look at tokenizers in their entirety; without it I can actually see the diff.

**Comment by RdoubleA** [comment]
> Want to express my uncertainty on the TokenEncoding naming. This seems a bit unintuitive since I'd imagine the encoding to refer to something like UTF8.

Yeah I agree I don't like the TokenEncoding naming either, but naming SP and TT as Tokenizer and everything else as ModelTokenizer is more confusing imo... there should be a clear distinction between a model tokenizer and the base tokenizer. Maybe we could call it `SentencePieceBaseTokenizer` and `TikTokenBaseTokenizer`

File: torchtune/data/tokenizers/_tiktoken.py
Comment (RdoubleA) [reply torchtune/data/tokenizers/_tiktoken.py:54]: this is only specific to tiktoken and will not be used elsewhere, correct? but I don't feel strongly either way

File: torchtune/data/tokenizers/_tiktoken.py
Comment (RdoubleA) [reply torchtune/data/tokenizers/_tiktoken.py:25]: I'm actually not sure why I didn't just move EOS and BOS handling to the TikToken class..

File: torchtune/models/llama2/_tokenizer.py
Comment (RdoubleA) [reply torchtune/models/llama2/_tokenizer.py:92]: Yeah I mainly surfaced the encode/decode in the model tokenizers for generation utilities. Otherwise we'd have to access the underlying model for encoding/decoding... which maybe is ok if we enforce a common class attribute name in the model tokenizer protocol.

File: torchtune/models/mistral/_tokenizer.py
Comment (joecummings) [inline torchtune/models/mistral/_tokenizer.py:149]: Mistral has special tokens, no?

File: torchtune/models/gemma/_tokenizer.py
Comment (joecummings) [inline torchtune/models/gemma/_tokenizer.py:122]: Gemma has special tokens, no?

File: torchtune/modules/tokenizers/_utils.py
Comment (joecummings) [inline torchtune/modules/tokenizers/_utils.py:171]: Is this used anywhere right now? Or just including it for future use? If the latter - while I do like it - maybe leave for when we need it.

Comment (RdoubleA) [reply torchtune/modules/tokenizers/_utils.py:171]: This is currently used in llama3 and phi3

**Comment by RdoubleA** [comment]
@joecummings huh I guess Gemma does have special tokens, but our current tokenizer does not use them. what do you think about either punting the upgrade to later or parsing the special tokens from the HF json for now and then adding them in tokenize_messages as appropriate later? Same question for Mistral

File: torchtune/models/llama2/_tokenizer.py
Comment (joecummings) [reply torchtune/models/llama2/_tokenizer.py]: Could we call it "decode_messages" on the `ModelTokenizer` and have it handle filtering out the special tokens?

File: torchtune/models/llama2/_tokenizer.py
Comment (ebsmothers) [inline torchtune/models/llama2/_tokenizer.py:56]: I don't like this being in the model-specific tokenizers. It's very much a hack to get `encode(a) + encode(b) = encode(a+b)` to hold for sentencepiece. But if I am a user copy-pasting tokenizers I am gonna have no idea what the hell is going on here

File: torchtune/models/gemma/_tokenizer.py
Comment (ebsmothers) [inline torchtune/models/gemma/_tokenizer.py:69]: Discussed a bit offline, but currently the encode/decode methods being defined in both Tokenizer and ModelTokenizer is confusing. Right now it's needed because we do call encode directly on ModelTokenizer in our generate recipe, but as @RdoubleA pointed out we should see if we can move everything onto tokenize_messages API (and deprecate encode/decode methods for ModelTokenizers) after #1106

File: torchtune/models/llama3/_tokenizer.py
Comment (ebsmothers) [inline torchtune/models/llama3/_tokenizer.py:36]: So I understand the purpose of this and it is nice that we can just directly load from HF. But I also don't love that (a) we are very tied to their specific tokenizer representation (and not just generic json) and (b) it kinda abstracts away the actual special tokens a bit (though I guess you do define them below). Another approach here would be to pass `special_tokens: Dict[str, int]` and leave it to the builder to take the tokenizer json path and unpack it. Tbh I don't have a strong opinion here though

Comment (joecummings) [reply torchtune/models/llama3/_tokenizer.py:36]: Ohh the second option sounds good. We will provide the function to load and build the special tokens via the HF file but allow people to pass in a dict if they want. Smart cookie.

File: torchtune/models/llama2/_tokenizer.py
Comment (joecummings) [reply torchtune/models/llama2/_tokenizer.py]: BUmp

Comment (RdoubleA) [reply torchtune/models/llama2/_tokenizer.py]: sorry I missed this. Yes I agree, will do that

File: recipes/eleuther_eval.py
Comment (RdoubleA) [reply recipes/eleuther_eval.py:45]: oh no

File: torchtune/datasets/_chat.py
Comment (RdoubleA) [reply torchtune/datasets/_chat.py:51]: yes, will revise

File: torchtune/models/llama2/_tokenizer.py
Comment (RdoubleA) [reply torchtune/models/llama2/_tokenizer.py]: actually, I removed `include_special` from decode here, so this no longer applies right? and the generate recipe still requires using `decode` from the `ModelTokenizer`

File: tests/torchtune/datasets/test_alpaca_dataset.py
Comment (ebsmothers) [inline tests/torchtune/datasets/test_alpaca_dataset.py:23]: Any concerns that by removing these tokenizers we are now not actually testing any tokenizers that have their open proper encode method?

File: tests/torchtune/modules/tokenizers/test_sentencepiece.py
Comment (ebsmothers) [inline tests/torchtune/modules/tokenizers/test_sentencepiece.py:21]: Why do we need to set this explicitly?

File: torchtune/models/mistral/_tokenizer.py
Comment (ebsmothers) [reply torchtune/models/mistral/_tokenizer.py:142]: I don't think so? https://huggingface.co/mistralai/Mistral-7B-v0.1/blob/main/tokenizer_config.json

File: torchtune/models/phi3/_tokenizer.py
Comment (ebsmothers) [inline torchtune/models/phi3/_tokenizer.py:52]: nit: what if I want to explicitly pass an empty dict (i.e. no special tokens)? I think this will override that, right? (Maybe this isn't a realistic case, I'm not super familiar with phi3 tokenizer usage)

File: torchtune/modules/tokenizers/_utils.py
Comment (ebsmothers) [inline torchtune/modules/tokenizers/_utils.py:32]: Maybe I'm being dense, I don't see `include_special` being used anywhere?

File: torchtune/modules/tokenizers/_utils.py
Comment (ebsmothers) [inline torchtune/modules/tokenizers/_utils.py:131]: (b) is no longer checked here. Personally I think it's OK to leave this in as we want to be explicit about what we're doing here, but would at least update this comment to be a bit more clear about the fact that this should be checked inside the base tokenizer's encode method

File: torchtune/models/gemma/_tokenizer.py
Comment (ebsmothers) [reply torchtune/models/gemma/_tokenizer.py:115]: Yeah what's the latest on this? It seems to me that it does have special tokens: https://huggingface.co/google/gemma-2b/blob/main/tokenizer_config.json

File: torchtune/models/llama3/_tokenizer.py
Comment (ebsmothers) [inline torchtune/models/llama3/_tokenizer.py:55]: Not a huge deal, but I feel like this isn't the most configurable way to do this. Like we claim to support custom special tokens, but they also need to be a superset of LLAMA3_SPECIAL_TOKENS, which is not immediately obvious and requires some digging around to figure out. I guess the obvious alternative here is something like `extra_special_tokens` which tbh I don't love either. Maybe just add an assert or something with an explicit error message about which keys must be present when using this tokenizer.

File: torchtune/modules/tokenizers/_sentencepiece.py
Comment (ebsmothers) [inline torchtune/modules/tokenizers/_sentencepiece.py:76]: The rationale here is also less clear now that we've split encode and tokenize_messages. Please make sure to document why we're doing this

**Review by ebsmothers — APPROVED** [review]
A handful more comments but no huge concerns from my side. Accepting to unblock

File: tests/torchtune/modules/tokenizers/test_sentencepiece.py
Comment (RdoubleA) [reply tests/torchtune/modules/tokenizers/test_sentencepiece.py:21]: because the test model actually does not encode whitespaces and breaks the test. Alternatively, I can update the test values

File: torchtune/models/phi3/_tokenizer.py
Comment (RdoubleA) [reply torchtune/models/phi3/_tokenizer.py:52]: It will override it, but I don't think this is a case we need to worry about. If there's no special tokens, then why use the Phi3Tokenizer? But, to be extra cautious, it's easy to just make this check for None explicitly.

File: torchtune/modules/tokenizers/_utils.py
Comment (RdoubleA) [reply torchtune/modules/tokenizers/_utils.py:32]: It's not. I may just remove it. It's not straightforward to enable including special tokens for SentencePiece, at least.

File: tests/torchtune/modules/tokenizers/test_sentencepiece.py
Comment (ebsmothers) [reply tests/torchtune/modules/tokenizers/test_sentencepiece.py:21]: Sorry, what I mean is: why is this not already figured out by `SentencePieceBaseTokenizer`? Linking to where this should happen

File: torchtune/modules/tokenizers/_sentencepiece.py
Comment (ebsmothers) [inline torchtune/modules/tokenizers/_sentencepiece.py:48]: Shouldn't the test tokenizer already have `encodes_whitespace` properly defined here?

File: tests/torchtune/models/phi3/test_phi3_tokenizer.py
Comment (RdoubleA) [reply tests/torchtune/models/phi3/test_phi3_tokenizer.py:16]: at least phi3 should be tested separately since the tokenize_messages is customized, but the others that simply use `tokenize_messages_no_special_tokens` there's not much difference. but it will be updated in the near future once they have their own special tokens (I need to look into Gemma more at least)

File: torchtune/models/gemma/_tokenizer.py
Comment (RdoubleA) [reply torchtune/models/gemma/_tokenizer.py:115]: these aren't used in any capacity in our current gemma implementation. Also I think we discussed elsewhere that the gemma and mistral special token updates will be a follow up: #1118

File: torchtune/models/mistral/_tokenizer.py
Comment (RdoubleA) [reply torchtune/models/mistral/_tokenizer.py:142]: not for our current implementation. Mistral v0.3 does.

File: docs/source/tutorials/chat.rst
Comment (joecummings) [inline docs/source/tutorials/chat.rst:117]: nit: this is a function ref now

File: torchtune/models/llama3/_tokenizer.py
Comment (joecummings) [inline torchtune/models/llama3/_tokenizer.py:43]: Why wouldn't we just default this to the LLAMA3_SPECIAL_TOKENS?

File: torchtune/models/mistral/__init__.py
Comment (joecummings) [inline torchtune/models/mistral/__init__.py:29]: This is in the `__all__`, but I don't see it updated in the model documentation. Therefore any links to this will not render and users won't be able to know what is going on with these tokenizers. Same applies to all model tokenizers.

File: torchtune/modules/tokenizers/__init__.py
Comment (joecummings) [inline torchtune/modules/tokenizers/__init__.py:22]: This needs to be documented.

**Review by joecummings — COMMENTED** [review]
This is amazing work, but I see a few things that definitely need to be documented before landing.

File: torchtune/models/llama3/_tokenizer.py
Comment (RdoubleA) [reply torchtune/models/llama3/_tokenizer.py:43]: mutable types - if we put dict/list as default here it will be the same dict for all instances of Llama3Tokenizer. If it's ever updated it will updated for all of them. Very edge case, but better to just be safe.

File: docs/source/tutorials/chat.rst
Comment (RdoubleA) [reply docs/source/tutorials/chat.rst:117]: ah shoot

File: torchtune/modules/tokenizers/_sentencepiece.py
Comment (RdoubleA) [reply torchtune/modules/tokenizers/_sentencepiece.py:48]: Yes it is defined, it just turned out to be False

File: torchtune/models/llama3/_tokenizer.py
Comment (RdoubleA) [reply torchtune/models/llama3/_tokenizer.py:55]: Yeah, I think in the ruckus your original assert checks got thrown out. let me add that back in

**Pull request merged**
